apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs

```diff
@@ -0,0 +1,245 @@
+    /// application for the current block height.
+    block: TxBin,
+    /// The current space utilized by protocol transactions.
+    protocol_txs: TxBin,
+    /// The current space utilized by DKG encrypted transactions.
+    encrypted_txs: TxBin,
+    /// The current space utilized by DKG decrypted transactions.
+    decrypted_txs: TxBin,
+}
+
+impl From<&RequestPrepareProposal>
+    for BlockSpaceAllocator<states::BuildingDecryptedTxBatch>
+{
+    #[inline]
+    fn from(req: &RequestPrepareProposal) -> Self {
+        let tendermint_max_block_space_in_bytes = req.max_tx_bytes as u64;
+        Self::init(tendermint_max_block_space_in_bytes)
+    }
+}
+
+impl BlockSpaceAllocator<states::BuildingDecryptedTxBatch> {
+    /// Construct a new [`BlockSpaceAllocator`], with an upper bound
+    /// on the max size of all txs in a block defined by Tendermint.
+    #[inline]
+    pub fn init(tendermint_max_block_space_in_bytes: u64) -> Self {
+        let max = tendermint_max_block_space_in_bytes;
+        Self {
+            _state: PhantomData,
+            block: TxBin::init(max),
+            protocol_txs: TxBin::default(),
+            encrypted_txs: TxBin::default(),
+            // decrypted txs can use as much space as needed; in practice,
+            // we'll only need, at most, the amount of space reserved for
+            // encrypted txs at the prev block height
+            decrypted_txs: TxBin::init(max),
+        }
+    }
+}
+
+impl<State> BlockSpaceAllocator<State> {
+    /// Return the amount of space left to initialize in all
+    /// [`TxBin`] instances.
+    ///
+    /// This is calculated based on the difference between the Tendermint
+    /// block space for a given round and the sum of the allotted space
+    /// to each [`TxBin`] instance in a [`BlockSpaceAllocator`].
+    #[inline]
+    fn uninitialized_space_in_bytes(&self) -> u64 {
+        let total_bin_space = self.protocol_txs.allotted_space_in_bytes
+            + self.encrypted_txs.allotted_space_in_bytes
+            + self.decrypted_txs.allotted_space_in_bytes;
+        self.block.allotted_space_in_bytes - total_bin_space
+    }
+
+    /// Claim all the space used by the [`TxBin`] instances
+    /// as block space.
+    #[inline]
+    fn claim_block_space(&mut self) {
+        let used_space = self.protocol_txs.occupied_space_in_bytes
+            + self.encrypted_txs.occupied_space_in_bytes
+            + self.decrypted_txs.occupied_space_in_bytes;
+
+        self.block.occupied_space_in_bytes = used_space;
+
+        self.decrypted_txs = TxBin::default();
+        self.protocol_txs = TxBin::default();
+        self.encrypted_txs = TxBin::default();
+    }
+}
+
+/// Allotted space for a batch of transactions of the same kind in some
+/// proposed block, measured in bytes.
+#[derive(Debug, Copy, Clone, Default)]
+struct TxBin {
+    /// The current space utilized by the batch of transactions.
+    occupied_space_in_bytes: u64,
```

Comment:
wrt. this PR it looks like `occupied_space_in_bytes` == `allocated_space_in_bytes`, is there a reason to use the term "occupied" here? (it would be easier imo to use the same terms wherever we can if there is no difference)

Reply:
sort of related to this comment - https://github.com/anoma/namada/pull/778#discussion_r1033597447

Reply:
well, I called this `used_space` before, but you suggested I changed this to `occupied_space` :joy:

Reply:
but yeah, `allocated_space` is better